### PR TITLE
Adding initial support for nodeinfo

### LINKED ISSUE #528: Expose node information following the nodeinfo convention
Following some discussion in Pleroma about reporting basic info about a server, there has been some interest for the [nodeinfo](https://github.com/jhass/nodeinfo/blob/master/schemas/2.0/example.json) standard. They seem to be already used by diaspora*, Friendica and Hubzilla, to name a few. 

While I know we already have reporting of such info, maybe it could help play nice with other server implementations at some point?

### PR DESCRIPTION
**Pull request by rigelk**
closes #528

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -13,10 +13,15 @@
       "es2016",
       "es2017"
     ],
+    "typeRoots": [
+      "node_modules/@types",
```
Comment (Chocobozzz) [inline tsconfig.json]: Why did you need this?

File: tsconfig.json [hunk 2]
```diff
@@ -13,10 +13,15 @@
...
+      "server/models"
+    ],
     "types": [
       "node",
       "chai-xml",
-      "chai-json-schema"
+      "chai-json-schema",
+      "nodeinfo"
```
Comment (Chocobozzz) [inline tsconfig.json]: And this?

File: server/controllers/static.ts [hunk 3]
```diff
@@ -95,6 +103,45 @@ async function getVideoCaption (req: express.Request, res: express.Response) {
...
+async function generateNodeinfo (req: express.Request, res: express.Response, next: express.NextFunction) {
+  const { totalVideos: totalVideos } = await VideoModel.getStats()
+  const { totalLocalVideoComments: totalLocalVideoComments } = await VideoCommentModel.getStats()
+  const { totalUsers } = await UserModel.getStats()
+
+  const json: HttpNodeinfoDiasporaSoftwareNsSchema20 = {
+    version: '2.0',
+    software: {
+      name: 'PeerTube',
+      version: '1.0.0-beta.9'
```
Comment (Chocobozzz) [inline server/controllers/static.ts]: Fetch the version from package.json (see config.ts controller)

File: server/controllers/static.ts [hunk 4]
```diff
@@ -70,6 +73,11 @@ staticRouter.get('/robots.txt', (req: express.Request, res: express.Response) =>
   return res.send(CONFIG.INSTANCE.ROBOTS)
 })
 
+// nodeinfo service
+staticRouter.use('/.well-known/nodeinfo',
+  asyncMiddleware(generateNodeinfo)
```
Comment (Chocobozzz) [inline server/controllers/static.ts]: Maybe we could add cache here

File: server/controllers/static.ts (hunk 3 above)
Comment (rigelk) [reply server/controllers/static.ts]: Oh. It's so simple it hurts :')

File: tsconfig.json (hunk 1 above)
Comment (rigelk) [reply tsconfig.json]: I thought defining explicitly the typeRoots would require to redefine the existing ones.

File: server/initializers/constants.ts [hunk 5]
```diff
@@ -574,6 +576,36 @@ export {
...
+  ms:           1,
+  second:       1000,
+  minute:       60000,
+  hour:         3600000,
+  day:          3600000 * 24,
+  week:         3600000 * 24 * 7,
+  month:        3600000 * 24 * 30
+}
+
+export function parseDuration (duration, defaultDuration) {
```
Comment (Chocobozzz) [inline server/initializers/constants.ts]: Miss type

File: server/initializers/constants.ts [hunk 6]
```diff
@@ -574,6 +576,36 @@ export {
...
+      let unit = split[2].replace(/s$/i,'').toLowerCase()
+      if (unit === 'm') {
+        unit = 'ms'
+      }
+
+      return (len || 1) * (timeTable[unit] || 0)
+    }
+  }
+
+  return defaultDuration
```
Comment (Chocobozzz) [inline server/initializers/constants.ts]: Maybe add an error log, this should not happen

File: server/controllers/static.ts [hunk 7]
```diff
@@ -95,6 +109,48 @@ async function getVideoCaption (req: express.Request, res: express.Response) {
...
 
+async function generateNodeinfo (req: express.Request, res: express.Response, next: express.NextFunction) {
+  const { totalVideos } = await VideoModel.getStats()
+  const { totalLocalVideoComments } = await VideoCommentModel.getStats()
+  const { totalUsers } = await UserModel.getStats()
+
+  const json: HttpNodeinfoDiasporaSoftwareNsSchema20 = {
+    version: '2.0',
+    software: {
+      name: 'PeerTube',
```
Comment (rhaamo) [inline server/controllers/static.ts]: https://github.com/jhass/nodeinfo/blob/master/schemas/2.0/schema.json#L35
The software->name element shouldn't have uppercase chars

**Comment by rhaamo** [comment]
I'm not sure to understand the routing, ```/.well-known/nodeinfo``` seems to returns the whole nodeinfo json ?

From https://github.com/jhass/nodeinfo/blob/master/PROTOCOL.md it should return a list of links to the actual nodeinfo and associated version

**Comment by rigelk** [comment]
@rhaamo I just realised that 30min ago… Should be fixed now.

File: server/controllers/static.ts (hunk 7 above)
Comment (rigelk) [reply server/controllers/static.ts]: Thanks! I missed that one.

**Comment by rigelk** [comment]
Update: the caching function has trouble caching the response due to the `content-type` of the response, which Express definitely doesn't like (it parses the profile URI as a charset…).

Right now, the cache makes further requests timeout as it cannot return a value that could not be saved, but is marked as cached.

I'll try fixing the cache function.

**Comment by rigelk** [comment]
Update: the caching function fails at caching the response due to its `content-type`, which Express parses wrong (it doesn't support `profile=` URIs apparently).
